### PR TITLE
Cache script_structure API responses

### PR DESCRIPTION
**Pull request by wjordan**
Followup to #8774 with one improvement and two fixes:
## Improvement:
- add public caching headers to `ApiController#script_structure`. Will not apply on levelbuilder server, where `Script.should_cache?` is `false`.

This allows client browsers as well as CloudFront/Varnish to cache the value of this static script structure for a configurable period of time, equivalent to the amount we currently cache Pegasus document pages in each environment.

The http cache configuration for this path already properly includes the current locale in the cache key. The user session cookie is also included (not ideal), so this will only currently be cached in the CDN/http-proxies for anonymous requests and not logged-in users. So this change will mainly be helpful to browser-side caching, however proxy cacheability can be improved in the future with a change to the http cache configuration.

This sets the stage for performant and scalable uses of the `script_structure` API moving forward (as outlined in my [comment](https://github.com/code-dot-org/code-dot-org/pull/8774#issuecomment-224408143)).
## Fixes:

(to allow the cached objects to invalidate correctly on levelbuilder when levels/stage content is updated)
- Add `I18n.backend.cache_key` helper method to include localization contents along with the current locale in the `Stage#summarize` cache key. This will invalidate the cached object whenever any I18n localized strings are modified.
- Add `script.cache_key` to the `Stage#summarize` cache key, because some of the contents depend on the stage's parent `Script` contents, and so the cached object should be invalidated whenever the script itself is updated.

### CODE REVIEW

File: dashboard/app/controllers/api_controller.rb [hunk 1]
```diff
@@ -256,6 +257,15 @@ def section_assessments
 
   private
 
+  ONE_HOUR = 3600
+
+  def configure_caching
+    return unless Script.should_cache?
+    max_age = DCDO.get('pegasus_document_max_age', ONE_HOUR * 4)
+    proxy_max_age = DCDO.get('pegasus_document_proxy_max_age', ONE_HOUR * 2)
+    response.headers['Cache-Control'] = "public,max-age=#{max_age},s-maxage=#{proxy_max_age}"
```
Comment (joshlory) [inline dashboard/app/controllers/api_controller.rb:266]: Can we add a versioning key to this response and cache it forever?

File: dashboard/config/initializers/fast_localization.rb [hunk 2]
```diff
@@ -13,3 +13,11 @@
...
   I18n.t 'hello'
 end
+
+# Provides a cache key based on the current locale and contents of all localized strings.
+module I18n::Backend::CacheKey
+  def cache_key
+    "#{I18n.locale}-#{translations.hash}"
+  end
+end
+I18n::Backend::Simple.include(I18n::Backend::CacheKey)
```
Comment (joshlory) [inline dashboard/config/initializers/fast_localization.rb:23]: Is this the backend we use for all environments?

**Comment by joshlory** [comment]
Ping — is this PR still active?

**Comment by wjordan** [comment]
this is low priority and a bit risky with the way it could affect subtle caching behavior, so i don't want to usher this into production until i can take the extra time to give it a thorough look again and do some manual spot-testing as it goes into staging/test/production. let me know if it's important for any related/upcoming work and i can try to return to it soon enough, otherwise this will sit for a while longer.

**Comment by ashercodeorg** [comment]
CC @ashercodeorg FYI @balderdash

**Pull request closed without merging**